### ISSUE TITLE
Discuss Proposed - 0012-safecoin-implementation

### ISSUE THREAD

**Issue by Viv-Rajkumar** [maidsafe/rfcs#61]

**Comment by benjaminbollen** [comment]
I believe that under the assumption that TP >= TS (there are always more primary chunks than sacrificial chunks)
```rust
if TS < TP {
    FD = MT - (TP - TS) + 1
} else {
    FD = MT + 1
}
```

can be simplified to

```rust FD = TS +1``` which is just a bit easier to reason about?

**Comment by Fraser999** [comment]
If that assumption is true then yes the code and comments preceding it can be simplified.  I'm not sure the assumption _is_ true however.

**Comment by benjaminbollen** [comment]
it can then equally be expanded to 

``` rust
if TS <= TP {
     FD = TS + 1
} else {
     FD = TS + 1
}```
which is simplified itself to
```rust
FD = TS + 1
```

**Comment by Fraser999** [comment]
That's also true, but I wanted to show how the bias could be affected in the future and I felt that leaving this form of the calculation helped explain that better.  However, I didn't realise that paragraph has been removed (https://github.com/dirvine/rfcs/commit/18033e81a1a4baeef231268eb6ce77369c8df38b) so I agree that it doesn't make sense to keep the more complex form here.

**Comment by dirvine** [comment]
Happy (as always) to have the simplified form if you wish to edit that @Fraser999

**Comment by robbo1** [comment]
Why have MT anyhow, the test defines which is the higher value.

Also the IF statement when reduced always produces FD = TS + 1

**Comment by maqi** [comment]
Not sure what will be the scope for this RFC, however I think there are couple of points need to be clarified in this RFC or others before any implementation work starts:
1, It is mentioned PmidNode will reply a wallet address when responding a get request.
    This requires vault executable to be started with user account info (via UI or some config file)
    It probably needs some work in vault and launcher to implement this

2, The Client Manager (or MaidManager in old name?) is defined to grant an account with initial credit.
    However, so far, the process of how a Client Account will be created is not explained in detail in this RFC (or any other?).

3, The detailed definition of the SafeCoin data structured has not been covered in this RFC, neither the explanation of the correspondent persona manager (I think when it is mentioned that ```This request is a POST message as a safecoin request``` ```if there is a safecoin available of the name result```, we are referring to a special personal SafeCoinManager handing it, however there is no separate section define it in the current RFC).


4, According to the RFC and white paper, ClientManager is in charge of a client's account (crediting or debting), DataManager is handling data request and trigger a farming attempt, SafeCoinManager is in charge of approval of farming attempt.
The RFC seems so far only explained what's happening in DM, but lack of info regarding what happens in CM and SCM.

5, As the safecoin is only generated through GET, what will be the benefit for pmid_node provide the storage?
Does the store_cost needs to be composed of two parts: safe_coin and storage_place ?
i.e. you are required not only provide the storage service to the network, but also owned some safe_coin, then you can store something to the network?

6, As there will be cache in routing layer, does it mean the farming procedure will be affected?

**Comment by dirvine** [comment]
Yes @maqi  some good points there. 

1. Yes it is expected the vault will have a config file with a minimum of two (optional) settings. These will be:
a. Wallet address (set by owner of vault)
b. Max disk space to use 

2. Yes we should add some pseudo code for this. 

3. We will detail this more, it's not clear a SafeCoinManager is required. It should be relatively simple addition for DataManager (to handle network types, like DNS, SafeCoin). Again needs detailed. 

4. Will be clear when 3 is detail. 

5. If a PmidNode does not have a chunk stored it won't get awarded any safecoin, so this one is easier :-) This is in fact an easy way to implement `rank` as the rank of a PmidNode is determined by the number of successful Get requests it can satisfy.

6. Yes this has always been a goal that cache speeds up data access and provides a natural ceiling for popular data. Cache will be number of chunks limited and not time.  The trade off is that data provided from cache does not earn safecoin as it cannot be validated and would be `game-able` 

Thanks @maqi

**Comment by Fraser999** [comment]
Thanks for your comment @robbo1

I believe this was the same point @benjaminbollen was making above and which I answered there.  The RFC has been updated (https://github.com/maidsafe/rfcs/pull/62) after discussion with @dirvine.

**Comment by robbo1** [comment]
Not quite.  Ben's was saying TS is always less than TP.   Mine was simply there is no need for MT and the logic error in the IF statement.  Its a mute point now though :-)

**Comment by maqi** [comment]
@dirvine : regarding the config file to be used for vault, does it being covered in this RFC, or shall we raise a separate one for it?

**Comment by robbo1** [comment]
Is this the place to ask a question and clarify the effects?

Specifically the Farm Request Calculation.

Am I correct in thinking that the effect of   (2 questions)

`6. If this result % farming divisor (modulo divides) yields zero then `
` .... ... If there is a safecoin available of the name result then`

will be that when the coin is scarce then this will typically result in no coin even if FR ~= 1  (FD ~= 1)

Additionally does this mitigate against an attack where if a particular chunk was rapidly got then the result is likely to be the same since various nodes are still the same?  That is very rapid GETting of a chunk would end up (if % == 0) requesting the same coin and the coin can only be given once?

**Comment by robbo1** [comment]
Looking at the code abstraction proposed, I am somewhat confused as to what is trying to be achieved with put_cost and storage_balance

I noticed that when doing a PUT the code abstraction says

            if store_cost() > storage_balance {
                Error::NotEnoughBalance;
            } else {
                storage_balance -= store_cost();
            }


Query:

 Should that not be a test to see if 
    storage_balance < #Puts_to_be_done 
and storage balance is updated 
    storage_balance -= #Puts_to_be_done;  

Otherwise the calculation when spending coin to buy storage space doesn't add up

I cannot see how the code in the RFC can be correct because once you plug reasonable figures in for FR, NC and group_size=32

say NC = 120, FR=.001  (1000 gets per coin attempt), you get store_cost=0.00375

BUT store_cost is defined as u64 not as a float ?????

And if doing a put then either storage_balance -= 0  (or 0.00375 if it should have been float)
and when spend a coin storage_balance += 266

If we took store_cost as float (& storage_balance) then really it is 266 squared puts per coin.

Is there a mistake???? Or can someone explain it please.  I am trying to put together some dynamic simulations and need to be sure if #puts per coins is 1/store_cost or that squared or something else.

Thnaks

**Comment by robbo1** [comment]
Looking at the code abstraction proposed, I am somewhat confused as to what is trying to be achieved with put_cost and storage_balance

I noticed that when doing a PUT the code abstraction says

            if store_cost() > storage_balance {
                Error::NotEnoughBalance;
            } else {
                storage_balance -= store_cost();
            }


Query:

 Should that not be a test to see if 
    storage_balance < #Puts_to_be_done 
and storage balance is updated 
    storage_balance -= #Puts_to_be_done;  

Otherwise the calculation when spending coin to buy storage space doesn't add up

I cannot see how the code in the RFC can be correct because once you plug reasonable figures in for FR, NC and group_size=32

say NC = 120, FR=.001  (1000 gets per coin attempt), you get store_cost=0.00375

BUT store_cost is defined as u64 not as a float ?????

And if doing a put then either storage_balance -= 0  (or 0.00375 if it should have been float)
and when spend a coin storage_balance += 266

If we took store_cost as float (& storage_balance) then really it is 266 squared puts per coin.

Is there a mistake???? Or can someone explain it please.  I am trying to put together some dynamic simulations and need to be sure if #puts per coins is 1/store_cost or that squared or something else.

Thnaks

**Comment by smacz42** [comment]
I have drafted up the reasoning behind a Put Incentive as an alternative form of App Reward payment that may address the Unresolved Question portion of this RFC [here](https://github.com/smacz42/Put_Incentive_Model). There is no code, only reasoning out the structure that such an implementation may use.
